core/src/main/java/org/elasticsearch/search/aggregations/pipeline/having/BucketSelectorPipelineAggregator.java

```diff
@@ -136,20 +141,83 @@ protected void doReadFrom(StreamInput in) throws IOException {
     public static class Factory extends PipelineAggregatorFactory {
 
         private Script script;
-        private GapPolicy gapPolicy;
+        private GapPolicy gapPolicy = GapPolicy.SKIP;
         private Map<String, String> bucketsPathsMap;
 
-        public Factory(String name, Map<String, String> bucketsPathsMap, Script script, GapPolicy gapPolicy) {
+        public Factory(String name, Map<String, String> bucketsPathsMap, Script script) {
             super(name, TYPE.name(), bucketsPathsMap.values().toArray(new String[bucketsPathsMap.size()]));
             this.bucketsPathsMap = bucketsPathsMap;
             this.script = script;
+        }
+
+        /**
+         * Sets the gap policy to use for this aggregation.
+         */
+        public void gapPolicy(GapPolicy gapPolicy) {
             this.gapPolicy = gapPolicy;
         }
 
+        /**
+         * Gets the gap policy to use for this aggregation.
+         */
+        public GapPolicy gapPolicy() {
+            return gapPolicy;
+        }
+
         @Override
         protected PipelineAggregator createInternal(Map<String, Object> metaData) throws IOException {
             return new BucketSelectorPipelineAggregator(name, bucketsPathsMap, script, gapPolicy, metaData);
         }
+
+        @Override
+        protected XContentBuilder internalXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.field(BucketScriptParser.BUCKETS_PATH.getPreferredName(), bucketsPathsMap);
+            builder.field(ScriptField.SCRIPT.getPreferredName(), script);
+            builder.field(BucketScriptParser.GAP_POLICY.getPreferredName(), gapPolicy.getName());
+            return builder;
+        }
+
+        @Override
+        protected boolean overrideBucketsPath() {
+            return true;
+        }
+
+        @Override
+        protected PipelineAggregatorFactory doReadFrom(String name, String[] bucketsPaths, StreamInput in) throws IOException {
+            Map<String, String> bucketsPathsMap = new HashMap<String, String>();
```

Comment:
bucketsPaths doesn't get used here as this implementation overrides bucketsPaths and uses its own parameter (a map instead of an array) so each bucketPath can be named and then referenced by that name in the script. 

In terms of the superclass reading stuff and then passing to the subclass I think it needs to be this as the superclass needs to read the stuff that it requires (mandatory options) to be passed into its constructor but then has to pass this stuff to the subclass since it doesn't know how to construct the instance of the subclass. It then has to set the optional parameters on the returned instance so they are set properly. I wouldn't want each subclass to have to separately read and set all the options which are generic to all pipeline aggregator factories.